(WIP) Skeleton for the metric data structure

afein opened the pull request:
I'm creating a data structure that consumes metrics from the cache and organizes them in a fashion that can be easily consumed by REST endpoints.

At the current state, only node metrics are processed and exposed, without any aggregation over cluster metrics.

@rjnagal @vishh @vmarmol 

File structure:
- types.go - contains the interface and types of the data structure
- impl.go - contains all methods of the RealCluster Implementation
- util.go - contains helper functions that will be reused extensively by the implementation

Next steps:

- unit tests and error handling for all functions on util.go
- implement updateContainer & addContainer
- implement updatePod & addPod
- implement GetNode, GetPod and GetContainer

Temporary concerns:
-Update should not be invoked during housekeeping (discussion pending)

rjnagal commented on api/repr/api.go:
```diff
@@ -0,0 +1,68 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package repr
```
Can you expand the name for readability. Here, and in all methods.

Does it stand for representation? You can also name it topology or schema.

rjnagal commented on api/repr/api.go:
```diff
@@ -0,0 +1,68 @@
+func (a *ReprApi) exportAllClusterMetrics(request *restful.Request, response *restful.Response) {
+	clinfo, _, _ := a.cluster.GetAllClusterData()
+
+	glog.Infof("Exported All Cluster Metrics")
```
This is going to print on every request. Probably make it glog.V(2).Infof?

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (rc *realCluster) GetAllClusterData() (*ClusterInfo, time.Time, error) {
+	/* Returns the entire ClusterInfo object */
```
I'd prefer a single comment style - " // "

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+)
+
+func NewCluster() Cluster {
+	var zero time.Time
```
Time should be zero by default. We can skip updating it and just fill in the cinfo below.

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+	var zero time.Time
+
+	// Invoke cache interface since the last timestamp
+	nodes := c.GetNodes(rc.timestamp, zero)
```
Inline zero as time.Time{}

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+	nodes := c.GetNodes(rc.timestamp, zero)
+	latest_time := rc.timestamp
+
+	for _, node := range nodes {
```
How do we take care of nodes that got deleted?

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+
+	node_name := node_container.Hostname
+
+	if _, ok := rc.Nodes[node_name]; !ok {
```
do we need the lock around lookup?

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+	return updateInfoType(&rc.Nodes[node_name].InfoType, node_container)
+}
+
+func (rc *realCluster) addNode(hostname string) *NodeInfo {
```
We don't check the return type? Do we need it to retutn the nodeInfo? Shall we return an error (eg if node exists).

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+	return &new_node
+}
+
+func (rc *realCluster) updatePod(pod *cache.PodElement) {
```
All Update*() should return error.

rjnagal commented on api/repr/impl.go:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	// Update the Cluster timestamp to the latest time found in the Update sequence
+	rc.lock.Lock()
```
It might be better if we are consistent with the other methods here and add an UpdateTime() call. The lock can then be subsumed there and we don't need to worry about addition of extra code after timestamp causing a deadlock.

rjnagal commented on api/repr/types.go:
```diff
@@ -0,0 +1,78 @@
+type InfoType struct {
+	Metrics map[string][]*MetricTimeseries
+	Labels  map[string]string
+	lock    *sync.RWMutex
```
Why do we split ClusterInfo from realCluster with just the timestamp at the top-level?
It's harder to see that lock here is protecting the timestamp in realCluster.

Can we push all fields of ClusterInfo into realCluster, or do we plan to reuse ClusterInfo somewhere else?

rjnagal commented:
ok to test

afein commented:
@rjnagal Addressed all comments in the new commit.

Note on locking: I will initially implement cluster-level locking, with plans to eventually support a more fine-grained update procedure

afein commented:
@rjnagal @vishh @vmarmol Added Unit tests for the schema skeleton, currently at 92.6% coverage.
The untested branches correspond to error handling/logging routines in the Update and updatePod methods.

afein commented:
To facilitate the code review process, I will submit the same content as independent pull requests of 50-200 lines of code. 

The PRs will be tagged as [Schema N], with N representing the number in the PR sequence.

vishh commented:
Closing this PR since this PR is being split up into multiple ones.
